packages/gatsby/src/utils/changed-pages.ts

```diff
@@ -0,0 +1,54 @@
+import { boundActionCreators } from "../redux/actions"
+const { deletePage, deleteComponentsDependencies } = boundActionCreators
+
+import { isEqualWith, IsEqualCustomizer } from "lodash"
+import { IGatsbyPage } from "../redux/types"
+
+export function deleteUntouchedPages(
+  currentPages: Map<string, IGatsbyPage>,
+  timestamp: number
+): string[] {
+  const deletedPages: string[] = []
+
+  // Delete pages that weren't updated when running createPages.
+  currentPages.forEach(page => {
+    if (
+      !page.isCreatedByStatefulCreatePages &&
+      page.updatedAt < timestamp &&
+      page.path !== `/404.html`
+    ) {
+      deleteComponentsDependencies([page.path])
+      deletePage(page)
+      deletedPages.push(page.path, `/page-data${page.path}`)
+    }
+  })
+  return deletedPages
+}
+
+export function findChangedPages(
```

Comment:
What happens to perf at a million pages?

Reply:
At 100k pages it's < 1s. There was an implementation of this in the inc build cli that was O(n^2), and which was the reason the 100k site initially took 4 hours to build! This one is much quicker.